### PR TITLE
[language] Refactor core metering logic

### PR DESCRIPTION
**Pull request by tzakian**
This PR updates the core gas metering logic to no longer do any parallel interpretation. 

The logic is pretty straightforward generally.  A macro was chosen for hooking the metering in to the interpreter since this will give us a way to turn off gas and all gas-related overhead later on when/if we want to measure the precise overhead of metering on the interpreter.

One particular thing to note is that native functions no longer have a constant as their gas cost -- they pull this from the on-chain gas schedule now (and using bogus constants for now as before). This means that native functions now take an additional argument. They can then lookup the correct costs in the table internally (this also means that a native can use multiple different costs defined on chain if need be).

In terms of perf improvements, there are some: 
```
With new gas metering:
  peer_to_peer            time:   [58.080 ms 58.582 ms 59.221 ms]

With old gas metering:
peer_to_peer            time:   [61.362 ms 61.697 ms 62.018 ms]
                        change: [+2.1514% +3.2931% +4.5251%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

### CODE REVIEW

**Comment by dariorussi** [comment]
could you explain where the perf numbers are coming from?
is that the bench for peer to peer? if that is the case when I run it I found out it showed big swing in results even with no change at all. Could you verify that? (maybe I will download this diff and do it myself ;) ). Could we try to run the test only with numbers for about 5K transactions and see the difference?
Also if the difference is only a 5% could we try to remove size checking to see if the perf problem was related to walking the objects?
Finally do you want help to do that? I can download this PR and try different things. Let me know

**Comment by tzakian** [comment]
* Perf numbers came from a devserver. The first on this PR, and the other numbers on `upstream/master`
* These are indeed for `peer_to_peer` (in `benchmarks`). I think this was running 5k as well?
* I'm in the midst of doing a bit of playing around with profiling, so will let you know on the size stuff. 
Would love to chat about this :) I'm a big fan of performance tuning. Might be a bit premature at the moment though (just as long as it's in the right direction for now mainly).

**Comment by dariorussi** [comment]
> * Perf numbers came from a devserver. The first on this PR, and the other numbers on `upstream/master`
> * These are indeed for `peer_to_peer` (in `benchmarks`). I think this was running 5k as well?
> * I'm in the midst of doing a bit of playing around with profiling, so will let you know on the size stuff.
>   Would love to chat about this :) I'm a big fan of performance tuning. Might be a bit premature at the moment though (just as long as it's in the right direction for now mainly).

yeah more than perf tuning I'd say it's perf understanding which is always a god thing ;) so happy you are looking at the profiler. I want to run peer_to_peer in benchmark few times on my devserver to see the results and see if it actually shows anything in the interpreter...
we'll talk more tomorrow

File: language/vm/src/gas_schedule.rs [hunk 1]
```diff
@@ -444,3 +463,25 @@ pub fn calculate_intrinsic_gas(
...
+    BYTEARRAY_CONCAT = 6,
+    LENGTH = 7,
+    EMPTY = 8,
+    BORROW = 9,
+    BORROW_MUT = 10,
+    PUSH_BACK = 11,
+    POP_BACK = 12,
+    DESTROY_EMPTY = 13,
+    SWAP = 14,
+    WRITE_TO_EVENT_STORE = 15,
```
Comment (dariorussi) [inline language/vm/src/gas_schedule.rs]: don't see save_account here, are we missing it or is there a good reason for it?

File: language/vm/vm-runtime/src/gas_meter.rs [hunk 2]
```diff
@@ -134,227 +139,30 @@ impl<'txn> GasMeter<'txn> {
...
-            self.consume_gas(instruction_gas)
+            let gas_cost = self.gas_schedule.instruction_cost(instruction_key as u8);
+            let gas = gas_cost.total().mul(memory_size);
+            self.consume_gas(gas)
         } else {
             Ok(())
         }
     }
 
-    /// Calculate the gas usage for an instruction taking into account the current stack state, and
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/gas_meter.rs]: I feel like crying at this change ;)

**Comment by tzakian** [comment]
Looking at the profiling (just on the `peer_to_peer` benchmark): it brought the cost of gas metering from 10% of transaction execution time to around 0.8% of the total transaction execution time.

File: language/vm/vm-runtime/vm-runtime-types/src/native_functions/dispatch.rs [hunk 3]
```diff
@@ -80,6 +83,12 @@ pub fn resolve_native_function(
     NATIVE_FUNCTION_MAP.get(module)?.get(function_name)
 }
 
+pub fn native_gas(table: &CostTable, key: NativeCostIndex, size: usize) -> GasUnits<GasCarrier> {
```
[the remark this answers is not in the document]
Comment (tzakian) [reply language/vm/vm-runtime/vm-runtime-types/src/native_functions/dispatch.rs]: Possibly? I think we could change `mul` to be a checked multiplication and have this return a result. Worth thinking about more, though that should be a different PR.

**Pull request closed without merging**